Add declarative CI config, parsing scripts, and tests

jmckenzie-dev opened the pull request:
Patch by jmckenzie; reviewed by XXXX for CASSANDRA-18731

## Reviewer Guide
Start with `README.md`. That'll give you the general structure and layout.

We're probably missing in things in both `pipelines.cfg` and `jobs.cfg`, and we could potentially use some more tests fleshed out in `.build/config/test/test_config.sh`.

jmckenzie-dev committed:
Add declarative CI config, parsing scripts, and tests

Patch by jmckenzie; reviewed by XXXX for CASSANDRA-YYYY

dcapwell commented on .build/config/jobs.cfg:
```diff
@@ -0,0 +1,183 @@
+    testlist:find test/unit -name '*Test.java' | sed "s;^test/unit/;;g" | sort
+    cmd:ant test ${ANT_TEST_OPTS}
+
+job:cdc
+    parent:unit
+    description:Runs test suites with CDC enabled
+
+job:compression
+    parent:unit
+    description:Runs test suites with SSTable compression enabled
+
+job:test-oa
+    parent:unit
+    description:Tests oa sstable format
+
+job:test-system-keyspace-directory
+    parent:unit
+    description:Execute unit tests with a system keyspaces directory configured
+
+job:test-tries
+    parent:unit
+    description:Execute unit tests with trie memtables
```
I strongly feel we should not define each like this and instead should push this into the oringal `unit` definition

```
job: unit
  parent:base
  test_cases:
    - cdc: [true, false]
    - compression: [true, false]
    - oa: [true, false]
    - system_keyspace_directory: [true, false]
    - tries: [true, false]
```

doing that is clear that these are just running `unit` with different configs.  Now the one question becomes how we "mix"... should tries run with compression enabled and cdc enabled?  Should we base this on the environment that can handle such cases?

jmckenzie-dev replied:
Hm. It'd certainly make things read more cleanly.

The parameterization bit of it and encoding of positional parameters within a config file format concerns me. There's no locality between the bool's here and what they represent, and then you end up with downstream parsers that also need to couple w/that positionality rather than being able to retain a declarative relationship between key/value.

dcapwell commented on .build/config/jobs.cfg:
```diff
@@ -0,0 +1,183 @@
+    memory:6
+    cpu:4
+    storage:100
+    env:ANT_TEST_OPTS="-Dno-build-test=true -Dtmp.dir=${TMP_DIR}"
```
nit: `-Dtmp.dir=${TMP_DIR}` is specific to Jenkins if I remember correctly, so don't think its needed here and might be best to live in the Jenkins runner?

jmckenzie-dev replied:
Good catch. I thought that briefly in passing when I pulled it in but forgot to make note. Will remove.

dcapwell commented on .build/config/jobs.cfg:
```diff
@@ -0,0 +1,183 @@
+job:base
+    memory:6
+    cpu:4
+    storage:100
```
100 what?

jmckenzie-dev replied:
Did you read the header on the file? Is clearly spelled out there:
```
#   The following parameters indicate the max allowable of a given resource
#       memory:N
#           integer:in GB
#       cpu:N
#           integer:logical cpu count
#       storage:N
#           integer:GB allowable used
```
my goal was to keep the parsing of the values as simple and clean as possible. Do you think / prefer having the unit there with it? No intention of ever supporting any different units (nor can I see a case in which we'd need to)

ekaterinadimitrova2 replied:
Maybe change to `storage_in_GB: ` ?

dcapwell commented on .build/config/jobs.cfg:
```diff
@@ -0,0 +1,183 @@
+# Base reference profile; add new parameters here to ensure all jobs have a reasonable default
+#-----------------------------------------------------------------------------
+job:base
+    memory:6
```
6 what?

jmckenzie-dev replied:
See above ☝️

dcapwell commented on .build/config/jobs.cfg:
```diff
@@ -0,0 +1,183 @@
+#-----------------------------------------------------------------------------
+# Base reference profile; add new parameters here to ensure all jobs have a reasonable default
+#-----------------------------------------------------------------------------
+job:base
```
what "image" do we need?

jmckenzie-dev replied:
I don't understand the question. What are you referring to here in terms of "image"? C* image? Docker image? OS? C*? Something else?

dcapwell commented on .build/config/jobs.cfg:
```diff
@@ -0,0 +1,183 @@
+job:dtest-offheap
+    parent:dtest
+    description:Python based multi-node distributed tests with off-heap memtables
+    env:DTEST_ARGS="--skip-resource-intensive-tests --keep-failed-test-dir"
+    env:DTEST_ARGS="--use-vnodes --num-tokens=${NUM_TOKENS} --use-off-heap-memtables --skip-resource-intensive-tests"
```
is this `append` semantic?  in `yaml` this `overrides` line 126

jmckenzie-dev replied:
Similar to above, this is spelled out in the documentation for the file:
```
#   env:{key=val,key=val,key=val}
#       string:Comma-delimited list of key value pairs to export in runtime env for test
#       NOTE: env parameters should *all* be evaluated from the root of the hierarchy. Child env params that
#               override the parent are acceptable but the totality of env should be a composite of each env:
#               from root on down in sequential order.
```

Now, diverging from .yaml semantics here: probably not a great idea. That said, the current paradigm (ability to inherit params from base) is somewhat predicated on the additive capabilities here (i.e. > 1 env: key / value can exist, append only). My preference is to keep the ability to have divergent env in a base job config vs. one declared below, but if you'd prefer the locality (up to and including completely removing the ability to derive from a base config), I can see the merits of that argument. Would make for a much uglier, longer, repetitive configuration file, down which road leads things like thousands of lines of text in our .circle configs.

So not my _preference_, but also not a hill I'm looking to die on.

dcapwell commented on .build/config/pipelines.cfg:
```diff
@@ -0,0 +1,78 @@
+# and sign off on this if using non-reference CI environments.
+#
+# Failure to do so will lead to commits being reverted.
+pipeline:pre-commit
```
do we really have this defined like this?  I kinda force everyone to run 100% of the tests... so a subset pre-commit feels wrong to me...

jmckenzie-dev replied:
> do we really have this defined like this

In circle, the config has "javaN_pre-commit_tests" and "javaN_separate_tests". Given the `long`, `vnode` vs. `novnode`, `large`, `burn`, etc suites, things that we don't expect to fail often or change often but we'd still like to run with some periodicity, in the interest of both saving on resources as well as making things simpler for folks running their own CI, we've inherited the "pre vs. post" designation for these suites.

If someone downstream just wants to run the full post-commit suite matrixed across all params (cdc vs. non, JDK versions, vnode vs. none, etc), I don't think anyone's _stopping_ them from doing so.

My personal opinion: we're targeting the point of dimishing returns in catching defects from where we drew the line on pre-commit vs. post-commit in terms of energy consumption, compute time, costs, etc. Especially if we end up with a cloud account for the project that committers can use to run tests, being efficient with our resource usage in this context really matters.

dcapwell reviewed (commented):
quick skim, the general idea LGTM

jmckenzie-dev committed:
Switch over to .yaml proper

This isn't ready yet; pushing up for feedback on structure.

- Combined .cfg files into .build/config/cassandra_ci.yaml
- added separate regions for pipelines, runtime env, params for multiplexing
- added timeouts and multiplexing values into the config
- updated .build/config/test/test_cassandra_ci.yaml to roughly match same structure

jmckenzie-dev commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,342 @@
+# both work in isolation and work when combined with other params.
+
+
+#-----------------------------------------------------------------------------
```
@dcapwell - could I get you to take another look at the new .yaml structure here?

Changes:
- pipelines
- jobs
- runtime_env (i.e. base params and env for collection of jobs)
- parameters (for matrixing on post-commit)
- added timeouts and multiplex values here

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,342 @@
+    resources: {cpu=4, memory=4g, storage=25g}
+    env:
+      ANT_TEST_OPTS: "-Dno-build-test=true"
+    testlist: find test/unit -name '*Test.java' | sed "s;^test/unit/;;g" | sort
```
this isn't generic... this is "how do I find unit tests" and not `jvm`....

I don't mind having reused object, yaml exposes an "anchor" that lets you reuse and extend; we used this a lot in circle ci yaml to avoid copy/paste

jmckenzie-dev replied:
Yep, I thought the same thing on this. We could either:
1. change type to `type: jvm-unit`, but that moves away from runtime env into execution details
2. remove the testlist from it and go the anchor route to pull things in

Think I'll check out the circle config and then refactor w/the 2nd route.

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,342 @@
+    env:
+      ANT_TEST_OPTS: "-Dno-build-test=true"
+    testlist: find test/unit -name '*Test.java' | sed "s;^test/unit/;;g" | sort
+    test_timeout: 480000
```
what is the unit?

jmckenzie-dev replied:
I hate these #'s in the build.xml with the fury of a thousand suns. I just pulled them from there to here; I'll make sure we're unitized on everything in the file. 👍

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,342 @@
+      ANT_TEST_OPTS: "-Dno-build-test=true"
+    testlist: find test/unit -name '*Test.java' | sed "s;^test/unit/;;g" | sort
+    test_timeout: 480000
+    multiplex: 500
```
what does this mean from an "environment" point of view?  this feels like a config to a specific job and not a generic thing

jmckenzie-dev replied:
Pulling out of env

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,342 @@
+    test_timeout: 900000
+    multiplex: 500
+    cmd: pytest ${PYTEST_OPTS} --cassandra-dir=${CASSANDRA_DIR} --keep-failed-test-dir ${DTEST_ARGS} ${TESTLIST}
+    env:
```
are there assumptions about ordering?  can I do ```env:
  FOO: "bar"
  BAR: "${FOO}"``` and have this setup `BAR=bar`?  

Or is the order of applying "undefined"?  I ask as `DTEST_ARGS` has a dependency on `NUM_TOKENS`, which makes this something we need to be explicit about defining

jmckenzie-dev replied:
From an ordering perspective, ISTM we can document / enforce / expect a "everything else needs to be resolved and parsed before you parse `cmd` and `env`.", and only allow those 2 to interpolate env variables.

There's some hand-waving around "some of these items become environment variables and other env var are declared explicitly in the env: section" I don't currently love. I'm not familiar enough w/yaml to know if there's an idiomatic solution to that or if this is something we should document and punt on.

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,342 @@
+#-----------------------------------------------------------------------------
+# JVM based single node tests
+#-----------------------------------------------------------------------------
+test_suites:
```
what's the diff between a `job` and a `test_suite`?  Can we just use one naming?

jmckenzie-dev replied:
Yep, for sure. Fixed.

I didn't do a _great_ job of setting expectations on this - was wanting to get some higher level "does this structure make more sense than the previous" kind of sniff testing. Appreciate the detail checking though!

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,342 @@
+  - job: unit
+    runtime_env: jvm
+    description: Runs unit tests with default configuration
+    cmd: ant testclasslist ${TESTLIST}
```
does this actually work with `ANT_TEST_OPTS` env?  This is the usage I see on trunk

```
$ grep -r ANT_TEST_OPTS .
./.build/run-tests.sh:        ant jar dtest-jar ${ANT_TEST_OPTS} -Dbuild.dir=${TMP_DIR}/cassandra-dtest-jars/build
./.build/run-tests.sh:    ant $_testlist_target -Dtest.classlistprefix="${_target_prefix}" -Dtest.classlistfile=<(echo "${testlist}") -Dtest.timeout="${_test_timeout}" ${ANT_TEST_OPTS} || echo "failed ${_target_prefix} ${_testlist_target}  ${split_chunk}"
./.build/run-tests.sh:  export ANT_TEST_OPTS="-Dno-build-test=true -Dtmp.dir=${TMP_DIR}"
./.build/run-tests.sh:      ant stress-build-test ${ANT_TEST_OPTS}
./.build/run-tests.sh:      ant $target ${ANT_TEST_OPTS} || echo "failed ${target} ${split_chunk}"
./.build/run-tests.sh:      ant fqltool-build-test ${ANT_TEST_OPTS}
./.build/run-tests.sh:      ant $target ${ANT_TEST_OPTS} || echo "failed ${target} ${split_chunk}"
./.build/run-tests.sh:      ant $target ${ANT_TEST_OPTS} -Dmaven.test.failure.ignore=true
./.build/run-tests.sh:      ant testclasslist -Dtest.classlistprefix=distributed -Dtest.timeout=$(_timeout_for "test.distributed.timeout") -Dtest.classlistfile=<(echo "${testlist}") ${ANT_TEST_OPTS} || echo "failed ${target} ${split_chunk}"
./.build/run-tests.sh:      ant testclasslist -Dtest.classlistprefix=distributed -Dtest.timeout=$(_timeout_for "test.distributed.timeout") -Dtest.classlistfile=<(echo "${testlist}") ${ANT_TEST_OPTS} || echo "failed ${target} ${split_chunk}"
^C
```

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,342 @@
+    description: Python based multi-node distributed tests with default configuration
+
+  - job: dtest-novnode
+    runtime_env: python
```
`type` and `runtime_env` are getting mixed....

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,342 @@
+# PARAMETERS
+#-----------------------------------------------------------------------------
+# Since JVM tests are generally quicker, we can afford to multiplex across more params
+jdk_parameters:
```
these names are super not clear and confusion... this also feels like we could simplify the "processing" of this file and leverage yaml anchors?


```
shared_objects:
  - &jdk_version
     - 11
     - 17

...
job: unit
with_jdks: *jdk_version
```

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,342 @@
+#-----------------------------------------------------------------------------
+# Commands that can be pulled and run in pipelines
+#-----------------------------------------------------------------------------
+commands:
```
I think I know where you copy from xD...

this is dead code

jmckenzie-dev replied:
Hah. Well, this was still in the circle config so I assumed it was something we relied on and could make use of. Very happy to delete this.

dcapwell reviewed (commented):
here is a similar, but think much simpler format

```
# this is not defined as part of the spec, this is just a "collection" of pointers used lower down
# no implementation should care about the existence of "runtimes"
runtimes:
  - &jdk_runtime
    resources: {cpu: 42, memory: 5g, storage: 25g}
jobs:
  - &job_unit
    name: unit
    runtime: *jdk_runtime
    env:
      ANT_TEST_OPTS: -Dno-build-test=true
      TYPE: unit
      FILTER: ""
    testlist: find "test/${TYPE}" -name "*Test.java" ${FILTER:-} | sed "s;^test/${TYPE}/;;" | sort
    run: .build/run-test.sh ${TYPE} ${SPLIT_DIR}/split

  - <<*job_unit
    name: jvm-dtest
    # why override name?  its because yaml anchors require you to define the full value, so would need another anchor if you want partial override of env
    env_override:
      TYPE: distributed
      FILTER: "| grep -v -e distributed/upgrade/ -e distributed/fuzz/"

pipelines:
  - name: pre-commit
    jobs:
      - unit
      - jvm-dtest
```

This defines 2 different primates: `job` and `pipeline`.  A pipeline is a simple list of jobs, and jobs can reference and extend other jobs to avoid copy/paste.

One complication not dealt with here is the "image" used... we want to run cross multiple JDK versions and have the same docker image defined...  this implies to me that we will want something like the following in `runtime`

```
image:
  template: "domain/path/${rhel}/jdk-${java}:latest"
  values:
    - rhel: 7
    - java: 11
    - java: 17
```

you define this list in the `runtime` level and can override it in pipeline if you want to only run a subset of jdk versions

jmckenzie-dev committed:
Anchor / alias / simplify structure

Ignore all changes in the branch / PR excepting the cassandra_ci.yaml file

jmckenzie-dev commented:
Ok @dcapwell, I've refactored things quite a bit, in some ways the way you were recommending, in other ways diverging. Commit w/changes to `.yaml` pushed up.

Not looking for fine-toothed comb pass yet; if you have 15 minutes to read through it and see if anything I've done now triggers an immune response, that should be enough for me to start digging into rewriting the config parsing scripts to use `yq` and pulling config from this file.

Instead of defining a "runtime", I went ahead and just aliased out some of the resource, timeout, and repeat params for re-use and convenience since enough subtle things vary that the notion of a "runtime" doesn't actually add much value in retrospect. Especially if most jobs are cleanly inheriting those values from the base anchor.

```
# Downstream test orchestration needs to use <= the following values when running tests.
small_executor: &small_executor {cpu: 4, memory: 1g, storage: 5g}
medium_executor: &medium_executor {cpu: 4, memory: 6g, storage: 25g}
large_executor: &large_executor {cpu: 4, memory: 16g, storage: 50g}

# Be opinionated and don't make people do ms <-> min conversions
timeout_1h: &16h_in_ms 57600000
timeout_15m: &15m_in_ms 900000
timeout_10m: &10m_in_ms 600000
timeout_8m: &8m_in_ms 480000

# Be opinionated; discourage bespoke repeat amounts on suites where we can.
repeat_default: &repeat_default 500
repeat_less: &repeat_less 100
repeat_tiny: &repeat_tiny 25
```

Further, I'm in favor of leaving out the image: * that you alluded to above. I don't think we should mandate OS / docker image / etc level constraints in this config file. If our requirements for pre-commit are:
1. Certain set of jobs
2. With certain resource constraints
3. On a specific JDK (oldest supported)
4. With a specific software environment (ant, JDK, python, env vars, etc)

But we _aren't_ requiring a specific distribution, docker image, or all JDK's, I don't think including those details in here adds value quite yet. The canonical CI will be on the ASF CI env where the OS, docker images, etc. are all very clearly defined, bespoke, and our 1-off source of truth. While we _could_ specify them in this file and update scripts and jenkins env to derive from here, I'm not sweating what form that takes day one (since it'd require unsticking some of the stuff in `.build/docker/*` as well)

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,307 @@
+  #
+  # Failure to do so can lead to commits being reverted.
+  - name: pre-commit
+    jdk:
+      - 11
```
given this is a list, are the jobs within the pipline all run with this as a config?  so this is equal to

```
for jdk in [11]:
  for job in jobs:
    job.apply(jdk=jdk)
```

jmckenzie-dev replied:
Right. The idea being impl just expects jdk for pipeline to be a list and enumerates over all members w/expectation of count >= 1

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,307 @@
+
+# There are some general environment variables devs will often need to override for parameterized runs.
+default_env_vars: &default_env_vars
+  ANT_HOME: /usr/share/ant
```
this can be iffy as it depends on the environment the tests are run on, but since this iteration wants to leave that out, then this kinda hard codes a very specific environment... feel we should delegate this to the runtime... "how" we document this is a requirement?  no idea!

jmckenzie-dev replied:
I split out to a required_env_vars and default_env_vars and promoted a bunch of the path env stuff to the latter and documented them. Should help make it incredibly clear which env vars are optional and which are required to be in compliance.

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,307 @@
+      # TYPE lines up with the various targets in build.xml for usage by the <testclasslist> target
+      TYPE: unit
+      TEST_FILTER: ""
+      TEST_TIMEOUT: *8m_in_ms
```
nit: I don't think we should define this here, its in `build.xml` so leave that as the source of truth... I can share code for how we made this generic in scripts in Jenkins and other environments.

jmckenzie-dev replied:
I'm on the fence of build.xml being the source of truth or the ci .yaml file being the source of truth tbh. build.xml is a pain in the butt.

That said, that's a bigger argument; maybe the right answer is to pull timeouts out of the .yaml and provide a reference build.xml timeout parser for a given job name as a helper function. I'll target that route instead.

michaelsembwever replied:
I am for build.xml being the lowest layer, and the context here.
See my comment wrt `java.default` and `java.supported`.

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,307 @@
+      TEST_FILTER: ""
+      TEST_TIMEOUT: *8m_in_ms
+    test_list: find "test/${TYPE}" -name "*Test.java" ${TEST_FILTER:-} | sed "s;^test/${TYPE}/;;" | sort
+    repeat: *repeat_default
```
I don't think this should be top level, `repeat` is not a top level thing IMO, it leverages primitives defined.  I would rework how repeat jobs are done

jmckenzie-dev replied:
So repeat jobs are an implementation detail at this point we're not specifying at this level. At this level we're saying "any time you repeat a test in this given job, you must do so N times". Hence the contract declaring the requirement here, much like the contract declaring resource constraints here even though the _usage_ of that is explicitly left to downstream consumers.

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,307 @@
+      CCM_HEAP_NEWSIZE: "512M"
+      CCM_CONFIG_DIR: ${DIST_DIR}/.ccm
+      NUM_TOKENS: "16"
+      TMPDIR: "$(mktemp -d /tmp/run-python-dtest.XXXXXX)"
```
I know you are trying to show an idea, but I would recommend that the commands be a script and anything that should be "tuned" is at this level and the rest treat them as internal details

jmckenzie-dev replied:
Nope, I'm not trying to show anything. I'm blindly copy/pasting from the `run-python-dtest.sh` script and moving on. Pulling this back to being just `"$(mktemp -d)"` as default_env_vars entry and leaving it up to implementations to override.

dcapwell reviewed (commented):
I think this is much better and liking it more!  Few high level notes

1) there is nothing special about `repeat` jobs, they are different ant task and shell scripts that leverage the same primitives; I don't think they should be a top level thing in a definition
2.1) an implementation has to figure out what to archive and make accessible, it might be good to think about how this is done.  If someone adds a new job, how does the implementation know what to archive?  
2.2) we *need* to archive xunit.xml, but its *desired* to archive the logs... how would we go about documenting this for an implementation?
3) I understand the desire to leave out the docker image, but this makes things complex for python-dtest... it also makes the implementations super coupled and easy to break as people make changes...  what if harry needs a brand new runtime, how would we "learn" about this without having to hand hold each imp?

dcapwell commented:
oh, one more feedback!  We have the ability to list the tests, but how do we know about the `cost` of the tests?  Right now Jenkins hard codes how many "splits" to have, and in CircleCI I try to get people to use my script that leverages a `cost` function to know how many resources are needed; it would be good to encode such a detail into this definition.

Here is the super complex `cost` function I use

```
unit: 20
jvm_dtest: 4
jvm_dtest_upgrade: 2
```

so how this works is `num splits = num_tests / cost`, this has been better at adapting to changes on the different branches and doesn't require we keep tuning CI... I want to forget about fixing CI!

smiklosovic commented on .build/config/README.md:
```diff
@@ -0,0 +1,125 @@
+Declarative Test Suite Configuration
+-------------------------------------------
+
+Pipeline and test wuite configurations are declarative so other CI implementations can build 
```
@jmckenzie-dev  wuite -> suite

jmckenzie-dev replied:
👍 Thanks for the catch. Fixed and will need to rewrite the file in general after we settle on final .yaml format.

jmckenzie-dev commented:
> oh, one more feedback! We have the ability to list the tests, but how do we know about the `cost` of the tests? Right now Jenkins hard codes how many "splits" to have, and in CircleCI I try to get people to use my script that leverages a `cost` function to know how many resources are needed; it would be good to encode such a detail into this definition.
> 
> Here is the super complex `cost` function I use
> 
> ```
> unit: 20
> jvm_dtest: 4
> jvm_dtest_upgrade: 2
> ```
> 
> so how this works is `num splits = num_tests / cost`, this has been better at adapting to changes on the different branches and doesn't require we keep tuning CI... I want to forget about fixing CI!

So... the higher the integer value for cost the... what, less splits you'll have right? Except that intersects with the number of tests?

What's the value compared between jobs supposed to communicate independently? i.e. a 20 on unit tests vs. 4 on python dtests means... unit tests are 5x cheaper to run? Or 5x smaller on average than a python dtest? Or do those numbers not actually mean anything independently of the count of tests on their respective jobs?

It's non-obvious to me why we'd have a _higher_ cost value for a unit test than a python dtest at face value (i.e. it's only clear once you look at how the calculation is used, and there's not an intuitive relationship between those #'s w/out both test count and that formula...)

Edit: As I think about this further, there _feels_ like there's redundancy between specified resources:, timeouts for tests, and this designation of cost. Is this something we can / should be calculating based on those factors + test count at runtime? i.e. provide a reference for that in the `.sh` files?

jmckenzie-dev commented:
> <cost talking>
Ok. So part of my reaction is the name `cost`; if we rename that to `split_size` I'm a lot more ok with it. But I'd think maybe we revise the calculation to something like:
`num_splits = (test_count / (split_size)) * $AGENT_COUNT_SCALAR`, where `AGENT_COUNT_SCALAR` is a user-provided fudge-factor downstream to indicate that, for a given test, they have N agents they can call up. That'd let people parameterize dialing up and down parallelization based on their unique env (i.e. if you have double the agents that can run a unit test at `medium_executor` vs. what you have available for `large_executor`, having some way to provide that to the scripts / reference system would be nice.

OR - we could just document that and let the users figure it out. But the num_splits = num_tests / hardcoded_value feels like it gives rise to a hardcoded value that is going to be env-specific.

So maybe... hm. We provide the `default_tests_per_split` (or whatever) for each job and document that folks may want to adjust that, and the higher they drive that number the more they can expect the suite to parallelize if agents are available?

jmckenzie-dev committed:
More review cycling on .yaml design

- Added repos: header section w/repo, branch, sha
- Removed timeout per job; kept one default
- Added default_splits: alias
- Split out default_env_vars and required_env_vars as top level]
- Added SPLIT_SIZE: to each job
- normalized all env var as all cap values

jmckenzie-dev committed:
Clean up TEST_LIST_FILE vs. TEST_SPLIT_FILE

- Add top level comment about impl required params (we're not populating the split files)
- Rename default_splits to default_split_num: to clarify
- tidied up TEST_FILE vs. TEST_SPLIT_FILE demarcation and usage
- unified jvm and dtest on test list default name

jmckenzie-dev commented:
Ok, highlights of these commits:

- Added a header that should make it clear we're not populating the `$TEST_SPLIT_FILE` and why, and how to do so
- Added some documentation on the flow of things to that header to clarify what's needed
- Added repo / branch / sha references at top of config for reproducibility and to provide the functionality around images you were concerned about (I think?) @dcapwell 
- Cleaned up jvm and dtest to use the same interim test file name
- tidied up some floating / varying cases (all upper should now be env var and requirements for what as env and what cmd should produce what env should be clear now)
- Pulled out the timeouts to your point David, but left in a default they all inherit from just in case a user adds a new job w/out specifying and there's no `build.xml` entry for it
- Cost calculation now a part of it via `default_split_num` that uses the `$SPLIT_SIZE` that we define on a per-job basis that users can also override. i.e. the higher the split size, the more tests we'll jam in there.
- Split out `default_env_vars` from `required_env_vars` and documented to try and make it clear what a downstream implementation should change and what they should leave alone.

Phew.

jmckenzie-dev committed:
Anchor top level python dtest url's and reference them from the job's env vars

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+repeat_tiny: &repeat_few 25
+
+# Default to at least one split
+default_split_num: &default_split_num let NUM_SPLITS=$(( $(wc -l < "$TEST_LIST_FILE") / $SPLIT_SIZE )); if [ "$NUM_SPLITS" -eq 0 ]; then NUM_SPLITS=1; fi
```
I honestly would drop this, think it makes sense to give a `hint` at cost of a single split, but doesn't make sense to generate it.

ekaterinadimitrova2 replied:
I do not understand... shouldn't we have this as default and give a chance to make the num_splits a variable instead of const? 1 by default, whatever you feel like otherwise... And you can plug your own way of doing splits, as it was mentioned earlier.

dcapwell commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+      TEST_FILTER: ""
+    TEST_LIST_FILE: ${DIST_DIR}/test_list.txt
+    test_list_cmd: find "test/${type}" -name "*Test.java" ${TEST_FILTER:-} | sed "s;^test/${type}/;;" | sort -R > ${TEST_LIST_FILE}
+    num_split_cmd: *default_split_num
```
not a fan, I recommend drop

ekaterinadimitrova2 replied:
And what do we do instead?

jmckenzie-dev committed:
wip

michaelsembwever commented on .build/config/README.md:
```diff
@@ -0,0 +1,125 @@
+* must run and pass on the lowest supported JDK before a committer merges any code
+### post-commit:
+* will run on the upstream ASF repo after a commit is merged, matrixed across more axes and including configurations expected to fail or diverge only rarely
+### nightly:
+* run nightly. Longer term, infra, very stable areas of code.
```
do we need this yet? 

rationale: if post-commit doesn't have a turn-around time requirement, and costs are not consumption based, then we can run everything post-commit.  we could also have sequential post-commit pipeline types, as well as sequential pipeline runs, ensuring feedback on the initial post-commit pipeline run is timely (which would essentially be similar to a nightly running repeatedly).  for example post-commit pipelines can have wait and idle times: how long to wait before starting and how long to be idle before running the next build (both which allow multiple commits to be collated into a build to avoid large pipeline build queues).

jmckenzie-dev replied:
Fair enough. Removed; kept it as an unopinionated option to consider during this process.

michaelsembwever commented on .build/config/assert.sh:
```diff
@@ -0,0 +1,266 @@
+# limitations under the License.
+#
+
+# Borrowed from https://github.com/torokmark/assert.sh/blob/main/assert.sh
```
will we need to add this to NOTICE.txt ?

michaelsembwever commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+  # Failure to do so can lead to commits being reverted.
+  - name: pre-commit
+    jdk:
+      - 11
```
can this be pulled from build.xml ? 
as it is already defined as `java.default`

jmckenzie-dev replied:
Maybe. This risks getting a little circular if I'm not careful... 🤔 

I've made a _bunch_ of changes I'm working on polishing up now including a `retrieve_value_from_xml() {` function that's inspired by the `_timeout_for` from the existing run scripts. Meaning we could define something like:
```
[cassandra_ci.yaml]
pipelines:
    - name: pre-commit
      jdk: java.default
    - name: post-commit
      jdk: java.supported
```
environments would have to know that the value in pipelines.[name].jdk is something they need to `retrieve_value_from_xml $BUILD_XML [value from jdk key]`. That'd let us indicate which JDK's a given job needs to run against and leave `build.xml` as the authority.

I'll go ahead and do that. I've already done that for the test.timeout values from build.xml.

michaelsembwever commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+      - 11
+    jobs:
+      - unit
+      - jvm-dtest
```
jvm-dtest-upgrade

jmckenzie-dev replied:
Oops. 😇

michaelsembwever commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+      - dtest-upgrade
+      - dtest-upgrade-large
+      - long-test
+      - cqlsh-test
```
fqltool
stress
burn
microbench

michaelsembwever commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+  - name: post-commit
+    jdk:
+      - 11
+      - 17
```
`java.supported` from build.xml ?

ekaterinadimitrova2 replied:
We added java.supported only in 5.0+

jmckenzie-dev replied:
This is going to start targeting trunk and backporting from there; should be trivial and low-pain to hard-code on a branch that isn't going to change in the future.

michaelsembwever commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+    jdk:
+      - 11
+    jobs:
+      - unit
```
this is `test` in the in-tree scripts

jmckenzie-dev replied:
So in build.xml we have:
```
    <property name="test.unit.src" value="${test.dir}/unit"/>
```
And the current .yaml for the job (not yet pushed; will push after local unit testing is done hopefully this week):
```
jobs:
  - &job_jvm_base
    name: unit
    resources: *medium_executor
    SPLIT_SIZE: 20
    env: &env_jvm_base
      <<: *env_default
      <<: *env_required
      ANT_TARGET: testclasslist
      ANT_TEST_OPTS: -Dno-build-test=true
      TEST_DIR: "test/unit"
      # This corresponds to the test.classlistprefix in build.xml, which is effectively the test/{TEST_CLASSLIST_PREFIX}.
      # While we could theoretically parse it out of TEST_DIR, or have the name overload or something, best to just keep it
      # decoupled and explicit for now.
      TEST_CLASSLIST_PREFIX: "unit"
      TEST_PACKAGE: "org.apache.cassandra"
      TEST_FILTER: ""
      # Corresponds to the tag value in build.xml containing the timeout for this test suite
      TEST_TIMEOUT: test.timeout
      TEST_SPLITTER: split_jvm_tests
      # We allow providing arbitrary arguments for the run: stage command
      RUN_ARGUMENTS: ""
    # These reference methods found in ci_functions.sh and other .sh files in .build/*.sh. None take arguments and all
    # rely on the environment as setup in either the execution environment, env:, or env_overrides:
    before:
      - setup_environment
      - pretest_setup
      - build_cassandra
    run:
      - run_jvm_tests
```
I strongly think "unit" is the right name for this given the test directory and classlist prefix from build.xml

michaelsembwever commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+      - 17
+    jobs:
+      - unit
+      - unit-cdc
```
test-cdc

michaelsembwever commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+      - dtest-upgrade
+      - dtest-upgrade-large
+      - long-test
+      - cqlsh-test
```
do we have an idea/plan of how to provide parameters to existing jobs, since that's all test-cdc, test-oa, test-system-keyspace-directory, test-tries, dtest-novnode, dtest-offheap are.

(even the duplication in build.xml should be parameterised in a more intelligent way)

jmckenzie-dev replied:
Yes. Resolved in current diff; will leave this here until that's pushed.

michaelsembwever commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+      - 17
+    jobs:
+      - stress-test
+      - fqltool-test
```
stress and fqltool are very small and fast.

jmckenzie-dev replied:
added

michaelsembwever commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+    jobs:
+      - stress-test
+      - fqltool-test
+      - test-burn
```
burn is important pre/post commit as breakages do happen. but should run for only a few second to ensure the code is working.

proper burn on longer runs would be appropriate for a nightly (or post-post-commit) pipeline

michaelsembwever commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+  TMPDIR: "$(mktemp -d)"
+  DIST_DIR: "${CASSANDRA_DIR}/build"
+  # Default to test.timeout as found in build.xml; should parse out of there in building local env and set this env var based on job
+  TEST_TIMEOUT: 480000
```
this varies for each test type

jmckenzie-dev replied:
Resolved in current diff. using `env.TEST_TIMEOUT` as k/v pair to indicate string of which timeout to pull from build.xml.

michaelsembwever commented on .build/config/ci_config_parser.sh:
```diff
@@ -0,0 +1,174 @@
+# This script relies on yq: https://github.com/mikefarah/yq
+# License: MIT: https://github.com/jmckenzie-dev/yq/blob/master/LICENSE
+# Plain binary install: wget https://github.com/mikefarah/yq/releases/download/${VERSION}/${BINARY} -O /usr/bin/yq &&\ chmod +x /usr/bin/yq
+# Brew install: "brew install yq"
```
can we do pre-conditions on each script like this: https://github.com/apache/cassandra/blob/trunk/.build/run-tests.sh#L29-L31

michaelsembwever commented on .build/config/jobs.yaml:
```diff
@@ -0,0 +1,16 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
not grokking the empty file, when all the jobs are in cassandra_ci.yaml

is this just for overrides/additions ?

jmckenzie-dev replied:
vestigial; removed.

michaelsembwever commented on .build/config/test/test_config.sh:
```diff
@@ -0,0 +1,117 @@
+#
+
+. ../ci_config_parser.sh
+. ../assert.sh
```
if this is the only place assert.sh is used, shouldn't we move it into test subfolder ?

jmckenzie-dev committed:
Overhaul .yaml to point to .sh

+ Split out public / private API in .sh
+ continue setting up steps on job_unit
+   - Update classlistprefix stuff to use what's defined by jobs in env
+ Start movement of .sh stuff from .yaml into .sh and call .sh from .yaml

michaelsembwever commented:
looking good.

i need to spend more time thinking about how this will integrate into the scripts and jenkinsfile.
test_list and splitting particularly comes to mind.
wondering if we can really merge this without it being taken in use somewhere…?

i am assuming this belongs in an architectural layer above the build system (build.xml) and the test scripts.
(the test scripts should be encapsulating their concerns).
but i see how this is itself its own utility and encapsulation.

ekaterinadimitrova2 commented on .build/config/README.md:
```diff
@@ -0,0 +1,125 @@
+* `storage:` Max allowable storage (in GB) allowable for a suite to access
+
+Jobs can be split up and parallelized in whatever manner best suits the environment in which they're
+orchestraed.
```
```suggestion
orchestrated.
```

ekaterinadimitrova2 commented on .build/config/README.md:
```diff
@@ -0,0 +1,125 @@
+   2. for each job:
+      1. `clear_job_params` to ensure nothing is left over from previous runs
+      2. `parse_job_params` to set up the params needed for the job
+      2. Write out the current job's params in whatever CI config format you're using in your 
```
```suggestion
      3. Write out the current job's params in whatever CI config format you're using in your 
```

ekaterinadimitrova2 commented on .build/config/README.md:
```diff
@@ -0,0 +1,125 @@
+As new entries are added to [pipelines.cfg](./cassandra_ci.yaml) and [jobs.cfg](./jobs.yaml), your 
+scripts should pick those up and integrate them into your configuration environment.
+
+Testing the in-tree config parsing scripts
+---------------------------------------------
+Currently testing is manual on the first addition of this declarative structure. As we integrate 
+it into our reference CI, we will integrate testing in as a new target.
+
+To run tests, execute [test_config.sh](./test/test_config.sh) from a terminal and inspect the 
+output.
```
I am not sure I understand this section. The title is about parsing, then we talk about new task and provide script name

jmckenzie-dev replied:
Obsolete now. More to come soon.

ekaterinadimitrova2 commented on .build/config/assert.sh:
```diff
@@ -0,0 +1,266 @@
+    BOLD=$(echo -en "\e[01m")
+fi
+
+log_header() {
```
I haven't looked in detail the script, but I like the idea. Seems neat.

ekaterinadimitrova2 commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Contains definitions of all pipelines and jobs (test suites) in Apache Cassandra's CI.
```
Maybe add some big title so it is easy to see where license stops, docs start?

ekaterinadimitrova2 commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+# IMPLEMENTATION REQUIRED PARAMETERS:
+#-----------------------------------------------------------------------------
+# We do not provide a mechanism to transform the contents of $TEST_LIST_FILE into $TEST_SPLIT_FILE. Implementations
+# must provide that mechanism and set that environment variable or "job->run:" operations will fail, unable to find a test split.
```
I would expect there is some default split that can be used if you do not have your own implementation?

jmckenzie-dev replied:
Providing it now based on blend of ASF CI + circle + some more cleverness w/shuf and md5 of build.xml for the SHA. More to come this week.

ekaterinadimitrova2 commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+# We do not provide a mechanism to transform the contents of $TEST_LIST_FILE into $TEST_SPLIT_FILE. Implementations
+# must provide that mechanism and set that environment variable or "job->run:" operations will fail, unable to find a test split.
+#
+# EXPECTED FLOW ON AN AGENT:
```
```suggestion
-----------------------------------------------------------------------------
# EXPECTED FLOW ON AN AGENT:
-----------------------------------------------------------------------------
```

ekaterinadimitrova2 commented on .build/config/cassandra_ci.yaml:
```diff
@@ -0,0 +1,355 @@
+# 1. Populate contents of $TEST_LIST_FILE for a given job using "job->test_list_cmd:" piped through "job->TEST_FILTER:"
+# 2. Split up $TEST_LIST_FILE using "job->num_split_cmd:"
+# 3. Populate $TEST_SPLIT_FILE with a given split (CI implementation specific)
+# 3. Execute "job->run:" to run the given $TEST_SPLIT_FILE
```
```suggestion
# 4. Execute "job->run:" to run the given $TEST_SPLIT_FILE
```

ekaterinadimitrova2 commented on .build/config/ci_config_parser.sh:
```diff
@@ -0,0 +1,174 @@
+KEY_TESTLIST_CMD="testlist"
+KEY_MEM="memory"
+KEY_CPU="cpu"
+KEY_STORAGE="storage"
```
probably we can add the unit as suffix?

ekaterinadimitrova2 commented on .build/config/functions.sh:
```diff
@@ -0,0 +1,53 @@
+# Confirm that a given variable exists
+# $1: Message to print on error
+# $2: Variable to check for definition
+check_argument() {
```
Maybe I would change the sh script name to Validations or something, functions doesn't ring a bell to me :-)

ekaterinadimitrova2 reviewed (commented):
Not a full review, but I skimmed through it and left tiny comments. It is going in the direction I expect. Well organized. From what I've seen, I am not a Jenkins expert, but I can't imagine this cannot be used there.

jmckenzie-dev committed:
Rewrite methods in the .sh file and tweak the .yaml a bit to use them

jmckenzie-dev committed:
Refactor yaml inheritance; get it passing validation

jmckenzie-dev committed:
Working through repeat jvm tests and starting repeat python tests

jmckenzie-dev committed:
Python repeat; various fixes and polish

- Remove vestigial env var setting at top of .sh
- Fix bad ref to dead ENV VAR for dtests
- unify test report generation into shared func
- make repeat count calculation shared func; call from jvm and python repeats
- fix bad code in test timeout calc in jvm line
- tidy up private section; jvm and python co-located
- make ant and pytest calls canonical and shared in single + repeat runs
- flip order on args to xml parsing to match grep

jmckenzie-dev committed:
Remove obsolete files and gut / update README.md

jmckenzie-dev committed:
review feedback

jmckenzie-dev committed:
[DO NOT MERGE] circleci hires for some timing testing
